### PR TITLE
Metrics deployment updates

### PR DESCRIPTION
**Pull request by astoycos**
Build on what @dave-tucker started

- Add some new metrics `bpf_programs`, `bpf_maps`, `bpf_links` Each of which report metadata for each individual entry on the system.
- Remove the program specific labels from `bpf_program_size_jitted_bytes_total`, `bpf_program_size_translated_bytes_total`, `bpf_program_mem_bytes_total`, and `bpf_program_verified_instructions_total`
- Add a default datasource of `http://localhost:9009/prometheus`
- Add a default dashboard `eBPF Subsystem Metrics`
- Update the README

### CODE REVIEW

File: bpf-metrics-exporter/README.md [hunk 1]
```diff
@@ -14,24 +14,32 @@ to correlate process IDs -> containers -> k8s pods.
...
-
-Labels:
-
-- `id`: The ID of the BPF program
-- `name`: The name of the BPF program
-- `type`: The type of the BPF program
-- `tag`: The tag of the BPF program
-- `gpl_compatible`: Whether the BPF program is GPL compatible
-- `load_time`: The time the BPF program was loaded
+- `bpf_programs`: Information on each loaded BPF Program
```
Comment (dave-tucker) [inline bpf-metrics-exporter/README.md:17]: If we're doing this, I'd rename this metric to `bpf_program_{N}_info`, where `N` is the program ID and the value is always 1. See: https://www.robustperception.io/exposing-the-software-version-to-prometheus/

File: bpf-metrics-exporter/README.md [hunk 2]
```diff
@@ -14,24 +14,32 @@ to correlate process IDs -> containers -> k8s pods.
...
+    - Labels:
+        - `id`: The ID of the BPF program
+        - `name`: The name of the BPF program
+        - `type`: The type of the BPF program as a readable string
+        - `tag`: The tag of the BPF program
+        - `gpl_compatible`: Whether the BPF program is GPL compatible
+        - `load_time`: The time the BPF program was loaded
+- `bpf_maps`: Information of each loaded BPF Map
```
Comment (dave-tucker) [inline bpf-metrics-exporter/README.md:25]: As above, please rename to `bpf_map_{N}_info` and add a `program_id` so we can associate it with a program

File: bpf-metrics-exporter/README.md [hunk 3]
```diff
@@ -14,24 +14,32 @@ to correlate process IDs -> containers -> k8s pods.
...
+    - Labels:
+        - `id`: The ID of the BPF map
+        - `name`: The name of the BPF map
+        - `type`: The type of the BPF map as an `u32` which corresponds to the following [kernel enumeration](https://elixir.bootlin.com/linux/v6.6.3/source/include/uapi/linux/bpf.h#L906)
+        - `key_size`: The key size in bytes for the BPF map
+        - `value_size`: The value size for the BPF map
+        - `max_entries`: The maximum number of entries for the BPF map.
+        - `flags`: Loadtime specific flags for the BPF map
+- `bpf_links`: Information on each of the loaded BPF Link
```
Comment (dave-tucker) [inline bpf-metrics-exporter/README.md:34]: As aboce, please rename to `bpf_link_{N}_info` and add a `program_id` label so we can associate it with a program

File: bpf-metrics-exporter/README.md [hunk 4]
```diff
@@ -14,24 +14,32 @@ to correlate process IDs -> containers -> k8s pods.
...
+    - Labels:
+        - `id`: The ID of the bpf Link
+        - `prog_id`: The Program ID of the BPF program which is using the Link.
+        - `type`: The BPF Link type as a `u32` which corresponds to the following [kernel enumeration](https://elixir.bootlin.com/linux/v6.6.3/source/include/uapi/linux/bpf.h#L1048)
+- `bpf_program_size_jitted_bytes_total`: The total size of all loaded BPF programs in bytes.
+- `bpf_program_size_translated_bytes_total`: The total size of all loaded BPF programs in bytes.
+- `bpf_program_mem_bytes_total`: The total amount of memory used by all loaded BPF programs in bytes.
+- `bpf_program_verified_instructions_total`: The total number of instructions in all loaded BPF programs.
```
Comment (dave-tucker) [inline bpf-metrics-exporter/README.md:42]: Please revert these changes to the metrics that were here.

File: bpf-metrics-exporter/README.md (hunk 2 above)
Comment (dave-tucker) [reply bpf-metrics-exporter/README.md:25]: Also, where is `bpf_map_memory_bytes`? We should have that as a metric for sure.

**Review by dave-tucker — CHANGES_REQUESTED** [review]
See comments. I kind of feel that so long as we end up with < 10 labels per time series we'll be okay - a lot of things there might not be useful and we can either not write them or drop them in the collector. 

I'd prefer that to having meta-metric objects to hold program/map/info metadata.

I know we'll be missing some debug information, but I think we should look at adding that into auditd and then ingesting through a Log pipeline or something similar.

File: bpf-metrics-exporter/README.md [hunk 5]
```diff
@@ -14,24 +14,32 @@ to correlate process IDs -> containers -> k8s pods.
...
+        - `gpl_compatible`: Whether the BPF program is GPL compatible
+        - `load_time`: The time the BPF program was loaded
+- `bpf_maps`: Information of each loaded BPF Map
+    - Labels:
+        - `id`: The ID of the BPF map
+        - `name`: The name of the BPF map
+        - `type`: The type of the BPF map as an `u32` which corresponds to the following [kernel enumeration](https://elixir.bootlin.com/linux/v6.6.3/source/include/uapi/linux/bpf.h#L906)
+        - `key_size`: The key size in bytes for the BPF map
+        - `value_size`: The value size for the BPF map
+        - `max_entries`: The maximum number of entries for the BPF map.
```
Comment (jotak) [inline bpf-metrics-exporter/README.md:32]: Like `load_time` above, moving them into their own metrics could make more sense. It would allow to compute stats (e.g. total/avg/... of sizes across all bpf programs)

File: bpf-metrics-exporter/README.md (hunk 2 above)
Comment (astoycos) [reply bpf-metrics-exporter/README.md:25]: We don't have `bpf_map_memory_bytes` yet from aya, I'll need to go back and add that

File: bpf-metrics-exporter/src/main.rs [hunk 6]
```diff
@@ -42,6 +43,25 @@ async fn main() -> anyhow::Result<()> {
     // Create a meter from the above MeterProvider.
     let meter = meter_provider.meter("bpf-metrics");
 
+    // TODO(astoycos) add nodename label to these gauges
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpf-metrics-exporter/src/main.rs:46]: +1 removed

File: bpf-metrics-exporter/src/main.rs [hunk 7]
```diff
@@ -69,55 +89,101 @@ async fn main() -> anyhow::Result<()> {
...
                             "load_time",
                             load_time.format("%Y-%m-%d %H:%M:%S").to_string(),
                         ),
+                        KeyValue::new("map_ids", format!("{map_ids:?}")),
                     ];
 
-                    observer.observe_u64(
-                        &bpf_program_size_jitted_bytes,
-                        jitted_bytes.into(),
-                        &labels,
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpf-metrics-exporter/src/main.rs:104]: Understood I added an abbreviated list of labels back to these that include `name, type and id`

File: bpf-metrics-exporter/README.md (hunk 2 above)
Comment (astoycos) [reply bpf-metrics-exporter/README.md:25]: Done on the rename, still need to add bpf_map_memory_bytes to aya

https://github.com/aya-rs/aya/issues/850

File: bpf-metrics-exporter/README.md (hunk 5 above)
Comment (astoycos) [reply bpf-metrics-exporter/README.md:34]: Done!

**Comment by astoycos** [comment]
I will follow this up with https://github.com/bpfman/bpfman/issues/858

File: bpf-metrics-exporter/README.md (hunk 4 above)
Comment (astoycos) [reply bpf-metrics-exporter/README.md]: Done

**Review by dave-tucker — APPROVED** [review]
LGTM. Apologies for the bike-shedding on this one 😆 
Let's get this in and we can continue to iterate on it until we're happy!
